sdk/monitor/Azure.Monitor.Ingestion/src/LogsIngestionClient.cs

```diff
@@ -266,141 +264,147 @@ public virtual async Task<Response<UploadLogsResult>> UploadAsync<T>(string rule
-                        true,
+                        async: true,
                         cancellationToken);
 
                     // Add the block to our task and commit lists
-                    runningTasks.Add(task);
+                    runningTasks.Add((task, batch.LogsList.Count));
 
                     // If we run out of workers
                     if (runningTasks.Count >= _maxWorkerCount)
                     {
                         // Wait for at least one of them to finish
-                        await Task.WhenAny(runningTasks).ConfigureAwait(false);
+                        await Task.WhenAny(runningTasks.Select(_ => _.CurrentTask)).ConfigureAwait(false);
                         // Clear any completed blocks from the task list
                         for (int i = 0; i < runningTasks.Count; i++)
                         {
-                            Task<BatchUpload> runningTask = runningTasks[i];
+                            Task<Response> runningTask = runningTasks[i].CurrentTask;
                             if (!runningTask.IsCompleted)
                             {
                                 continue;
                             }
-
-                            // Before removing runningTask - check if it has any errors and add to error list
-                            if (runningTask.Result.Error != null)
-                                errors.Add(runningTask.Result.Error);
-
+                            // If current task has an exception, log the exception and add number of logs in this task to failed logs
+                            if (runningTask.Exception != null)
+                            {
+                                exceptions = AddException(
+                                    exceptions,
+                                    runningTask.Exception);
+                                logsFailed += runningTasks[i].LogsCount;
+                            }
+                            // If current task returned a response that was not a success, log the exception and add number of logs in this task to failed logs
+                            if (runningTask.Result.Status != 204)
+                            {
+                                exceptions = AddException(
+                                    exceptions,
+                                    new RequestFailedException(runningTask.Result));
+                                logsFailed += runningTasks[i].LogsCount;
+                            }
                             // Remove completed task from task list
                             runningTasks.RemoveAt(i);
                             i--;
                         }
                     }
                 }
+                catch (Exception ex)
+                {
+                    // If we have an error, add error from response into exceptions list which represents the errors from all the batches
+                    AddException(
```

Comment:
I don't think this is necessary as we will loop over all the tasks anyway. We can probably combine this try/catch with the one that starts on line 336.